src/librustdoc/clean/types.rs

```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-crate struct Constant {
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, did: DefId },
```

Comment:
This one uses `print_inlined_const` and is generated from `clean/inline.rs` whereas `TyConst` simply uses `to_string()` from its expression.

Reply:
Right, sure, but why do you need it to be separate? Could you store this instead wherever you store a `TyConst`?

Reply:
Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. :laughing: 

Do you want me to add a FIXME so we can do it in another PR?

Reply:
> Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. laughing

Can't you just store the `DefId` instead of storing the whole `ty::Const`?

Reply:
The `ty::Const` comes from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/subst/enum.GenericArgKind.html so I wonder if `tcx.type_of(def_id)` will actually allow me to get it without too much troubles.

The easiest way seem to simply remove the clean impl for `ty::Const` and instead take it by value into the enum. However, it has a lifetime which is what we don't want to do here if we want to avoid a super big diff. I already added a FIXME, so unless there is a way to get the `ty::Const` from the `DefId`, I think it's better to keep it this way here and then move the `ty::Const` directly into the enum in another PR (and with all the lifetime changes it'll bring).

However, if you know a way to get the `ty::Const`, I'd be more than happy to do it here directly. I just couldn't find a way to get it though...

Reply:
@GuillaumeGomez why did you mark this as resolved? I vaguely remember we talked about this in DMs but I don't remember what we decided - I still feel like TyConst could be combined into `Extern` by only storing the DefId.

Reply:
I guess my question more broadly is why you need the `ty::Const` for `TyConst` when you don't need it for `Extern`.

Reply:
Because `ty::Const` is a generic, not a type and can't be retrived easily just from its `DefId`. Which is why you said on zulip that it was fine to keep as is.

Reply:
Thanks! That was it :)